### ISSUE TITLE
Retrieving BLOB data while having problems with special characters

### ISSUE THREAD

**Issue by mikeytroo** [oracle/node-oracledb#1108]
Hi guys, First I'll have to say I'm sorry if this isn't the right place to ask this question but I'm hoping I can receive some help since I'm running out of ideas. (Also I'm a newbie)
Node.js version: 8.15.0- x64
node-oracledb version: 2.3.0
Instant Client version: 19.3.0.0-x64
Oracle Database version: 11.2.0.4.0
DB Character sets:
- NLS_CHARACTERSET: EE8ISO8859P2
- NLS_NCHAR_CHARACTERSET: AL16UTF16

Basically I have to retrieve BLOB data from DB but I'm having troubles reading EXIF text of BLOB data/image. 
When trying to look at it I see special characters...
![2019-05-29_23-05](https://user-images.githubusercontent.com/5327760/58584592-ddb5fa80-8256-11e9-875f-322970c09ac6.png)

After that I do the following with node-oracle
```javascript 
execution_options.outFormat = oracledb.OBJECT;
execution_options.fetchInfo = {"DKM_IMG_DATA":{type:oracledb.BUFFER},"MUF_IMG_DATA":{type:oracledb.BUFFER}};
execution_options.fetchArraySize = 10000;
execution_options.resultSet = true;
try {
  let sql_query ="SELECT MUF.muf000_id "MUF_ID",
    MUF.muf000_n1 "MUF_N1",
    MUF.dkm001_content "DKM_IMG_DATA",
    FROM MUF0 MUF WHERE MUF.muf000_id =:id";
  let poolName = (process.env.ORA_DB_ENV == 'production') ? configDB.production.prodPool.poolAlias:configDB.staging.stagPool.poolAlias;
  dbconnection = await oracledb.getConnection(poolName);
  let result = await dbconnection.execute(sql_query, bind_parameters, execution_options);
  // Result set operation fetching as chunks/row
  let rows_return = [];
  let row;
  row = await result.resultSet.getRow();
  while (row) {
     rows_return.push(row);
     row = await result.resultSet.getRow();
   }
```
After I receive this i try the following where I try to check buffer for EXIF data but I always just receive size and ModifySize. I think exif-parser expects UTF-8.
```javascript 
const exif = require('exif-parser');
let chardet_result = chardet.detect(bufferObjects[i].DKM_IMG_DATA);
console.log(chardet_result); // prints out windows-1252
let parser = exif.create(bufferObjects[i].DKM_IMG_DATA);
let img_exif = parser.parse();
console.log(JSON.stringify(img_exif, null, 2)); // prints out nothing but size and modify date
fs.writeFileSync(`${tst_path}${bufferObjects[i].MUF_N1}`,bufferObjects[i].DKM_IMG_DATA, "binary", function (er) {
                    console.error(err);
});
```
I can create file/image just fine but none of this images have any metadata for reading its EXIF.

I'm wondering is there anyway to determine encoding conversion with node-oracle ?
Any help would be much appreciated.

**Comment by anthony-tuininga** [comment]
Can you provide a simple example as plain code? Its a little hard to follow what you are trying to do. Thanks!

**Comment by dmcghan** [comment]
@mikeytroo The tab that you're using in SQL Developer is for text, but you're trying to see an image, no? What do you get if you use the image tab?

> I'm wondering is there anyway to determine encoding conversion with node-oracle ?

If it's a BLOB, there is no encoding - it's just binary. It's up to you to know what's in there.

**Comment by mikeytroo** [comment]
When I use the Image tab I can the images just fine.

**Comment by dmcghan** [comment]
@mikeytroo Are you using this library?
https://github.com/gomfunkel/node-exif#node-exif

If so, it says:
> With node-exif you can extract Exif metadata from images (JPEG).

Don't use that library (unless you want the metadata). Just write the binary data you fetch from the database back to a file with the right extension.

**Comment by mikeytroo** [comment]
So far I've used exif-parser. 
So far I'm not writing files but I did try it and file can be written just fine but image metadata(EXIF) is missing. I'm trying to collect GPS data.

**Comment by dmcghan** [comment]
So your goal is to get at the metadata... Maybe try the other library? It says one of the properties is `gps`. But of course, you'd only be able to get that metadata if it was included in the file. Are you sure it was?

**Comment by mikeytroo** [comment]
Yes I'm sure, but It's true that wasn't my work. 
What I don't get is shouldn't I see the EXIF text normally when looking at TEXT tab on BLOB field ?
And since I can just write files back on disk just normally I don't get where is this metadata/EXIF data. When looking at images metadata, I can only see modify date and size.

**Comment by mikeytroo** [comment]
and once again sorry for the inconvenience.

**Comment by dmcghan** [comment]
> Yes I'm sure, but It's true that wasn't my work.

How can you be sure if it wasn't your work? Why not download a file and try viewing it in an application that knows how to work with Exif so you know whether or not the data is there?

> What I don't get is shouldn't I see the EXIF text normally when looking at TEXT tab on BLOB field ?

SQL Dev doesn't have built-in Exif support. It basically attempts a best guess as to your text encoding and shows that. It's really not all that different from opening the image with a text editor. Have you tried that? You need to use apps that understand how to work with Exif.

> It seems like embedded EXIF text is in different encoding...

Kinda. You can learn more about Exif encoding here:
https://stackoverflow.com/questions/1821515/how-is-exif-info-encoded

> That BLOB can seen just fine but it's metadata is missing/in different encoding.

The BLOB is the unchanged binary file that was uploaded.

Try ignoring Node.js for a moment and use SQL Dev to download an existing Exif file to your OS. Then use an app that understands Exif to try to find your metadata. If you can't find it, try this: take an Exif file that does have your metadata and upload that with SQL Dev and download it again. Is it still there? It should be. Then test that file in Node.js.

**Comment by mikeytroo** [comment]
I tried using SQL Dev to download BLOB/image and read it's metadata, no luck. I can save img just fine and view it but there is no metadata/exif.
I'll try to clear this things out with my coworker ASAP.
I have limited role inside Database because it's not ours, it's customers (my user can't insert/update/delete records).

Dan thanks for your help, I'm sorry if this question wasn't 100% related to node-oracle as I first thought.
